src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Simon Hudon
+-/
+
+import algebra.group_power
+import control.uliftable
+import control.monad.basic
+
+import data.bitvec.basic
+import data.list.basic
+import data.set.intervals.basic
+import data.stream.basic
+import data.fin
+
+import tactic.cache
+import tactic.interactive
+import tactic.norm_num
+
+import system.io
+import system.random
+
+/-!
+# Rand Monad and Random Class
+
+This module provides tools for formulating computations guided by randomness and for
+defining objects that can be created randomly.
+
+## Main definitions
+  * `rand` monad for computations guided by randomness;
+  * `random` class for objects that can be generated randomly;
+    * `random` to generate one object;
+    * `random_r` to generate one object inside a range;
+    * `random_series` to generate an infinite series of objects;
+    * `random_series_r` to generate an infinite series of objects inside a range;
+  * `io.mk_generator` to create a new random number generator;
+  * `io.run_rand` to run a randomized computation inside the `io` monad;
+  * `tactic.run_rand` to run a randomized computation inside the `tactic` monad
+
+## Local notation
+
+ * `i .. j` : `Icc i j`, the set of values between `i` and `j` inclusively;
+
+## Tags
+
+random monad io
+
+## References
+
+  * Similar library in Haskell: https://hackage.haskell.org/package/MonadRandom
+
+-/
+
+open list io applicative
+
+universes u v w
+
+/-- A monad to generate random objects using the generator type `g` -/
+@[reducible]
+def rand_g (g : Type) (α : Type u) : Type u := state (ulift.{u} g) α
+/-- A monad to generate random objects using the generator type `std_gen` -/
+@[reducible]
+def rand := rand_g std_gen
+
+instance (g : Type) : uliftable (rand_g.{u} g) (rand_g.{v} g) :=
+@state_t.uliftable' _ _ _ _ _ (equiv.ulift.trans.{u u u u u} equiv.ulift.symm)
+
+open ulift (hiding inhabited)
+
+/-- Generate one more `ℕ` -/
+def rand_g.next {g : Type} [random_gen g] : rand_g g ℕ :=
+⟨ prod.map id up ∘ random_gen.next ∘ down ⟩
+
+local infix ` .. `:41 := set.Icc
+
+open stream
+
+/-- `bounded_random α` gives us machinery to generate values of type `α` between certain bounds -/
+class bounded_random (α : Type u) [preorder α] :=
+(random_r : Π g [random_gen g] (x y : α),
+              (x ≤ y) → rand_g g (x .. y))
+
+/-- `random α` gives us machinery to generate values of type `α` -/
+class random (α : Type u) [preorder α] extends bounded_random α :=
```

Comment:
`random` shouldn't depend on `preorder`.  Most types have a reasonable notion of random value, but only few are ordered (in a canonical way).

Reply:
`bounded_random` does need `preorder`. I'd like one class to have both `random_r` and `random`. The reverse inclusion doesn't work because you can't have a uniform distribution on `nat`